### ISSUE TITLE
Uploader doesn't seems to work without CGM data

### ISSUE THREAD

**Issue by RalfCGM** [pazaan/600SeriesAndroidUploader#108]
#### Your Environment

* Uploader Version Number: 0.4.0
* Your Android Phone Model Name ): Moto X Play
* Android Version Number : 6.0.1
* Network Connection at the time the issue occurred (Wi-Fi, Mobile): -
* (if relevant) Nightscout website version : 0.9.2

#### Brief Explanation of Issue
No data from pump seems to be transferred to phone, while the connection to the pump is made: "Connected to Contour Next Link on channel 26". 

#### Uploader Error Code
No error codes mentioned, just: "never" updated.

#### Steps to Reproduce (for bugs)

1. Follow steps to install exactly. Including the Carelink logins and data upload to the carelink site.
2. Install software version 0.4.0 on phone.
3. Connect CNL with USB OTG cable.
4. Wait till: "Connected tot Contour Next Linkon channel 26". 
5. Observe that no data seems to be transferred. 

The only problem I could think of, is dat I don't use medtronic sensors. So in my situation there's no update of the BG values. When i touch "registered devices" there's: ".... and get a reading from the pump". But my 640g doens't have any BG readings. I just want the pumpdata (IOB, battery, reservoir levels) to transfer to nightscout. 

#### Severity Score
2 - would be really nice to have it working.

**Comment by volkerrichert** [comment]
I don't know if this has ever been tested and possible. Do you have connected the CNL and the pump? It has to be a device in the list of registered devices

**Comment by RalfCGM** [comment]
Thanks for you reply!

It's all connected together. It also says that the pump is connected on port 26, but my pump isn't in the list of registered devices, I suspect because it does not get BG readings.

As a workaround I've tried version 0.3 with the old carelink registration. That gave me a registered pump in the list on the nsuploader, but still no iob displayed on the app.  After installing 0.4 it didn't remember the registered pump.

Op 17 mrt. 2017 9:00 a.m. schreef volkerrichert <notifications@github.com>:

I don't know if this has ever been tested and possible. Do you have connected the CNL and the pump? It has to be a device in the list of registered devices

**Comment by Pogman** [comment]
The uploader will not record any data without a valid SGV and a SGV event time.

Branch: Master  File: MedtronicCnlIntentService.java

```javascript
                    boolean cancelTransaction = true;
                    if (pumpRecord.getSgv() != 0) {
                        // Check that the record doesn't already exist before committing
                        RealmResults<PumpStatusEvent> checkExistingRecords = activePump.getPumpHistory()
                                .where()
                                .equalTo("eventDate", pumpRecord.getEventDate())
                                .equalTo("sgv", pumpRecord.getSgv())
                                .findAll();

                        // There should be the 1 record we've already added in this transaction.
                        if (checkExistingRecords.size() <= 1) {
                            realm.commitTransaction();
                            cancelTransaction = false;
                        }

                        // Tell the Main Activity we have new data
                        sendMessage(Constants.ACTION_REFRESH_DATA);
                    }

                    if (cancelTransaction) {
                        realm.cancelTransaction();
                    }

**Comment by RalfCGM** [comment]
Thanks, that's clear.

It's such a pity I really can't work with you on the software, since my profession is technical HVAC engineer...

Are there maybe any possibilities to make this an enhancement, and make the uploader independent of the SGV's?


Op 17 mrt. 2017 14:08 schreef Pogman <notifications@github.com>:

The uploader will not record any data without a valid SGV and a SGV event time.

Branch: Master File: MedtronicCnlIntentService.java

                    boolean cancelTransaction = true;
                    if (pumpRecord.getSgv() != 0) {
                        // Check that the record doesn't already exist before committing
                        RealmResults<PumpStatusEvent> checkExistingRecords = activePump.getPumpHistory()
                                .where()
                                .equalTo("eventDate", pumpRecord.getEventDate())
                                .equalTo("sgv", pumpRecord.getSgv())
                                .findAll();

                        // There should be the 1 record we've already added in this transaction.
                        if (checkExistingRecords.size() <= 1) {
                            realm.commitTransaction();
                            cancelTransaction = false;
                        }

                        // Tell the Main Activity we have new data
                        sendMessage(Constants.ACTION_REFRESH_DATA);
                    }

                    if (cancelTransaction) {
                        realm.cancelTransaction();
                    }

**Comment by jrussell88** [comment]
I have the same issue, as I'm using a Freestyle Libre rather than Medtronic's Enlite sensors.
I did a fingerstick test on the Next Link 2.4 - had to disconnect from the phone to do this - in the hope that would enable it, but it didn't change. 

The app shows:
- "Connected to Contour Next Link on channel 23"
- Registered devices shows no registered devices

It's a show stopper if you don't use Enlite sensors.

**Comment by jrussell88** [comment]
Remains an issue in 0.5.0

**Comment by volkerrichert** [comment]
Yeap. Which data can be collected?

- time of pump
- battery level

Let's collect them
________________________________

**Comment by RalfCGM** [comment]
Indeed, I've also tried the 0.5.0, but data doesn't transfer.

Looks like there's still some dependancy on Bg values.


Op 17 jul. 2017 11:02 schreef volkerrichert <notifications@github.com>:
Yeap. Which data can be collected?

- time of pump
- battery level

Let's collect them
________________________________

**Comment by William1444** [comment]
👍

**Comment by jrussell88** [comment]
Arguably this is an error, not an enhancement, even if by design. 

Medtronic's upload works with or without Enlite sensors but this uploader doesn't work without Enlite sensors, which must include a lot of 640G users given the popularity of other methods of measuring BG.

**Comment by pazaan** [comment]
Sure, but this is in the context of Nightscout. I'm saying I'm not sure how useful a real-time remote monitoring system is when you're getting up to do a BG anyway. I understand that it's useful, but still not the majority of the use cases (for the purposes of prioritisation)

**Comment by Bal00** [comment]
We're using 640G with a mixture of Enlites and Freestyle Libre. We upload the Libre to Nightscout using xDrip+ and one of the versions of Libre Alarm. Getting pump data up to NS while using a different CGM would be great (IOB would be awesome alone!) :-)

Thanks for all your efforts Lennart and team!!

**Comment by Pogman** [comment]
The work on this is progress as it's needed for other enhancements too so that data gets sent to NS even when the sensor is wonky or in it's warm-up phase.

**Comment by Bal00** [comment]
Thanks @Pogman, I'll have a go at building and testing that :-)

**Comment by jrussell88** [comment]
@Bal00 How did the test go?

**Comment by th122** [comment]
I'm bitten by this, too - I received my 640G yesterday, but delivery of the CGM seems to have been postponed indefinitely. Guess I'm not the only one there, though.

- I haven't been able to get the upload to Carelink running as of yet (The best I got from all my attempts with different setups (hw/sw) was an upload of the meter, but it kept failing to discover the pump through the meter). Still trying there. Hence, no data from Carelink.
- The python tool by @pazaan [decoding-contour-next-link](https://github.com/pazaan/decoding-contour-next-link) does a terrific job: It's connecting to the pump without any hassles (Yay! At this stage of things that's a true relief to know!), and displays IOB, TBR and supplies level (not the fingerstick BG though, it seems).
- This uploader (Version 0.51) connects to the meter, discovers the pump through it and connects, and then complains about SGV not being available (too sad, it's right about that). Unfortunately, that's it. No active insulin, no fingerstick BG, no supply levels. (Even though the release note of 0.51 mentions the fingerstick BG, it doesn't get pulled into the uploader)
- Managed to get the preliminaries set up to compile and install the branch @Pogman referenced a few notes above. It needed to uninstall the 0.51 Version (of course), and installed itself instead. Unfortunately, it keeps crashing, though. (Android 7.0 device with API 24)

**Comment by Pogman** [comment]
@th122 sorry your having troubles. The "treatments" branch *should* be working well enough without crashes but maybe there is some sort of issue with it... I can't really support it as the current WIP is way beyond that branch now and that was not a release branch but a stepping stone to full pump history pulling and management.

Maybe some more info about what the crash is doing/reporting would help us help you. If you have set up a bugfender a/c you can access the program logs there.

**Comment by th122** [comment]
@Pogman Thanks for looking into this! Before blaming anythin else, I'd suspect the error to be on my side, especially since both branches (`pazaan:master` and your `ready branch for testing`) are crashing. 
It's the first time I'm using that Android SDK (Android Studio on Mac in particular) for anything other than some adb commandline fu. My Android developping experience starts out somewhere around 0. I found the process of getting the branch to compile at all surprisingly well automated and pain-free (well, nearly :), including the automated installation of the sdk versions determined as needed. I did agree to the suggestions to upgrade gradle as well - not knowing if I should leave it or not, and everything seemed to proceed. 
I did get (and add) a bugdefender key, but the bugdefender website doesn't show any logs from it. Maybe that's narrowing the problem down to some faulty integration on my side? I still have the Event Log of the  process:

```
02.11.2017
22:52	Filesystem Case-Sensitivity Mismatch
			The project seems to be located on a case-sensitive file system.
			This does not match the IDE setting (controlled by property "idea.case.sensitive.fs")
			More details. (show balloon)
22:53	Gradle sync started
22:53	Gradle sync failed: Failed to find target with hash string 'android-25' in: /Volumes/Users/theuser/Library/Android/sdk
			Consult IDE log for more details (Help | Show Log) (35s 669ms)
22:57	Gradle sync started
22:57	Gradle sync failed: Failed to find Build Tools revision 25.0.2
			Consult IDE log for more details (Help | Show Log) (3s 623ms)
22:57	Gradle sync started
22:58	Project setup started
22:59	Gradle sync finished in 1m 50s 344ms
22:59	Executing tasks: [:app:generateDebugSources, :app:generateDebugAndroidTestSources, :app:mockableAndroidJar, :app:prepareDebugUnitTestDependencies]
23:00	Gradle build finished in 31s 329ms
23:18	Executing tasks: [clean, :app:assembleDebug]
23:22	Gradle build finished in 3m 19s 795ms
23:30	Executing tasks: [:app:assembleDebug]
23:30	Gradle build finished with 1 error(s) in 163ms
23:30	Build APK(s): Errors while building APK. You can find the errors in the 'Messages' view.
23:33	Executing tasks: [:app:assembleDebug]
23:36	Gradle build finished in 2m 48s 524ms
23:36	Failed to finalize session : INSTALL_FAILED_UPDATE_INCOMPATIBLE: Package info.nightscout.android signatures do not match the previously installed version; ignoring!
23:36	Instant Run performed a full build and install since
			the installation on the device does not match the local build on disk.
			(Don't show again)
23:46	Executing tasks: [:app:assembleDebug]
23:46	Gradle build finished in 3s 20ms
```
... Sometimes semi-random clicking around rewards some experience: Found the inconspicious button to display the run log. Looks like I may have made an error with the crashlytics integration? (I did enable the plugin, and it added ` app/fabric.properties` as documented. Contrary to the `README.md`, the key was added to `app/src/main/AndroidManifest.xml` instead of the non-existing `app/src/AndroidManifest.xml`

```
11/02 23:46:40: Launching app
$ adb install-multiple -r -t [...]
Split APKs installed
$ adb shell am start -n "info.nightscout.android/info.nightscout.android.SplashActivity" -a android.intent.action.MAIN -c android.intent.category.LAUNCHER
Connected to process 19063 on device motorola-moto_g__5_-Z...
Capturing and displaying logcat messages from application. This behavior can be disabled in the "Logcat output" section of the "Debugger" settings page.
I/InstantRun: starting instant run server: is main process
E/CrashlyticsCore: .
E/CrashlyticsCore: .     |  | 
E/CrashlyticsCore: .     |  |
E/CrashlyticsCore: .     |  |
E/CrashlyticsCore: .   \ |  | /
E/CrashlyticsCore: .    \    /
E/CrashlyticsCore: .     \  /
E/CrashlyticsCore: .      \/
E/CrashlyticsCore: .
E/CrashlyticsCore: This app relies on Crashlytics. Please sign up for access at https://fabric.io/sign_up,
                   install an Android build tool and ask a team member to invite you to this app's organization.
E/CrashlyticsCore: .
E/CrashlyticsCore: .      /\
E/CrashlyticsCore: .     /  \
E/CrashlyticsCore: .    /    \
E/CrashlyticsCore: .   / |  | \
E/CrashlyticsCore: .     |  |
E/CrashlyticsCore: .     |  |
E/CrashlyticsCore: .     |  |
E/CrashlyticsCore: .
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
                  Process: info.nightscout.android, PID: 19063
                  java.lang.RuntimeException: Unable to create application info.nightscout.android.UploaderApplication: io.fabric.sdk.android.services.concurrency.UnmetDependencyException: This app relies on Crashlytics. Please sign up for access at https://fabric.io/sign_up,
                  install an Android build tool and ask a team member to invite you to this app's organization.
                      at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5405)
                      at android.app.ActivityThread.-wrap2(ActivityThread.java)
                      at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1541)
                      at android.os.Handler.dispatchMessage(Handler.java:102)
                      at android.os.Looper.loop(Looper.java:154)
                      at android.app.ActivityThread.main(ActivityThread.java:6123)
                      at java.lang.reflect.Method.invoke(Native Method)
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:867)
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:757)
                   Caused by: io.fabric.sdk.android.services.concurrency.UnmetDependencyException: This app relies on Crashlytics. Please sign up for access at https://fabric.io/sign_up,
                  install an Android build tool and ask a team member to invite you to this app's organization.
                      at com.crashlytics.android.core.CrashlyticsCore.onPreExecute(CrashlyticsCore.java:234)
                      at com.crashlytics.android.core.CrashlyticsCore.onPreExecute(CrashlyticsCore.java:207)
                      at io.fabric.sdk.android.InitializationTask.onPreExecute(InitializationTask.java:44)
                      at io.fabric.sdk.android.services.concurrency.AsyncTask.executeOnExecutor(AsyncTask.java:611)
                      at io.fabric.sdk.android.services.concurrency.PriorityAsyncTask.executeOnExecutor(PriorityAsyncTask.java:43)
                      at io.fabric.sdk.android.Kit.initialize(Kit.java:69)
                      at io.fabric.sdk.android.Fabric.initializeKits(Fabric.java:440)
                      at io.fabric.sdk.android.Fabric.init(Fabric.java:384)
                      at io.fabric.sdk.android.Fabric.setFabric(Fabric.java:342)
                      at io.fabric.sdk.android.Fabric.with(Fabric.java:313)
                      at info.nightscout.android.UploaderApplication.onCreate(UploaderApplication.java:32)
                      at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1046)
                      at android.app.ActivityThread.handleBindApplication(ActivityThread.java:5402)
                      	... 8 more
D/NetworkSecurityConfig: No Network Security Config specified, using platform default
Application terminated.
```
Any idea on how I should proceed?

**Comment by Pogman** [comment]
@th122 your missing the Crashlytics dependency. You will need to set up an account and install there plugin into AS to get going. There should be directions for doing this on MAC on their site. Been a while since I set it up so details are hazey... https://fabric.io/kits/android/crashlytics

**Comment by th122** [comment]
I think I'm giving up. I went through the Crashlytics install a second time (as guided on the Crashlytics web page when calling the dashboard there), it basically confirmed that everything was present, the Crashlytics "window" in the corner of Android Studio told me to "Launch yout app, once launched we will verify Crashlytics is configured for your app", and upon rebuild, reinstall and relaunch, it just kept barfing as quoted before
```
11-04 00:09:29.897 16225-16225/info.nightscout.android E/CrashlyticsCore: This app relies on Crashlytics. Please sign up for access at https://fabric.io/sign_up,
                                                                          install an Android build tool and ask a team member to invite you to this app's organization.
```
(Yes, I did sign up to Crashlytics and it did install the `io.fabric.ApiKey` when called from the newly integrated menu. I'm at the point where both Crashlytics 3-step webpage and the fabric menu tell me to "just" _Run the app_ . Too bad it won't run.)

**Comment by Pogman** [comment]
An alternative to get you going: open "640gAndroidUploader\app\src\main\java\info\nightscout\android\UploaderApplication.java" and delete the following code and it should build without invoking Crashlytics

```
        if (prefs.getBoolean(getString(R.string.preferences_enable_crashlytics), true)) {
            Fabric.with(this, new Crashlytics());
        }
        if (prefs.getBoolean(getString(R.string.preferences_enable_answers), true)) {
            Fabric.with(this, new Answers(), new Crashlytics());
        }
```

**Comment by th122** [comment]
Thanks - that did the trick, it does compile and run.
Complains about SGV being unavailable after connecting to the pump. Neither active Insulin nor BG get detected. I'll take a closer look tomorrow (oops, later today).

**Comment by nsewar01** [comment]
Just trying to upload my 640g results to NS after receiving the OTG cable this morning.
I am using Libre to monitor blood sugar and then manually inputting these values into pump to bolus purposes. Does that not mean that the values shuld be in the pump? or are these entries not SGV values?
The NS uploader V 0.5.1 reports the following on my Samsung S6:
>connecting to CNL
>Connecting on chanel 20 RSSI 63%
> SGV unavailable from pump
This seems similar to other users experience on this page. 
Unfortunately I do not have any knowledge of programming so am unable to follow the discussions as to the possible cause of this problem.
My CNL results are able to be uploaded to the carelink website without problem.
Any help/sugestions would be appreciated.

**Comment by Pogman** [comment]
@nsewar01 SGV = Sensor Glucose Value and Nightscout is cloud service for graphing these sensor values as read from the pump via the uploader.

Saying that, the next version of the uploader will work without SGV and allow you to send BG values to NS as well as all other info that's on the pump.

Check here or on FB for announcements when it's ready for release.

**Comment by nsewar01** [comment]
@pogmam Thanks for your quick reply. 
I was hoping to run "autotune" on its own as I dont have a loop set up yet. However,  it needs all the readings from my pump  in order to work. I look forward to the next itteration of the uploader programme here.
Many thanks

**Comment by Pogman** [comment]
closing feature available in v0.6+
